### PR TITLE
#1002 refactor TkSlf4j

### PR DESCRIPTION
**Pull request by pal548**
Actually it is my first opensource PR, and more like a training of Yegor's OOP style. I'll be grateful for some critics

https://github.com/yegor256/takes/issues/1002

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1003](https://github.com/yegor256/takes/issues/1003) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1003](https://github.com/yegor256/takes/issues/1003) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/C3FFK3YAY/c61900b3-9ece-4f0b-9ccd-0faebccfebb3); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by fanifieiev** [comment]
@pal548 please put the link to the issue in the description

**Comment by fanifieiev** [comment]
@pal548 The build fails due to a lot of checkstyle issues.
Youn first need to build the project on your side using `mvn clean install -Pqulice` to make sure the build  does not fail.
Please fix it and submit new commit.

**Comment by fanifieiev** [comment]
@paulodamaso Why was this PR assigned for REV. The underlying [1002](https://github.com/yegor256/takes/issues/1002) has not been accepted yet, but this PR is accepted for review.
Is it correct behavior of `0crat `?

**Comment by fanifieiev** [comment]
@0crat wait for ARC

**Comment by 0crat** [comment]
> @0crat wait for ARC ([here](https://github.com/yegor256/takes/issues/1003#issuecomment-561707273))

@fanifieiev It's a code review job [#1003](https://github.com/yegor256/takes/issues/1003), can't put it on hold

**Comment by fanifieiev** [comment]
@pal548 We do not usually review until build is successful.
Please have a look at the build detail [travic build](https://travis-ci.org/yegor256/takes/builds/620849708?utm_source=github_status&utm_medium=notification).

**Comment by paulodamaso** [comment]
@pal548 Also, take a look at [How to Contribute](https://github.com/yegor256/takes#how-to-contribute) section in README.md, it explains how to run the build locally using the same checks we use in our CI engines

**Comment by pal548** [comment]
Sorry guys, I thought I ran -Pqulice...

**Comment by fanifieiev** [comment]
@pal548 Once again, your build fails because there are a lot of checkstyle violations.

`[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[93]: Variable 'log' should be declared final. (FinalLocalVariableCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[99]: Parameters should start on a new line (BracketsStructureCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[100]: Indentation (28) must be same or less than previous line (12), or bigger by exactly 4 (CascadeIndentationCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[101]: Closing bracket should be on a new line (BracketsStructureCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[105]: Parameters should start on a new line (BracketsStructureCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[106]: Indentation (28) must be same or less than previous line (12), or bigger by exactly 4 (CascadeIndentationCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[107]: Closing bracket should be on a new line (BracketsStructureCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[112]: Missing a Javadoc comment. (JavadocTypeCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[113]: Missing a Javadoc comment. (JavadocMethodCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[116]: Missing a Javadoc comment. (JavadocTypeCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[117]: Missing a Javadoc comment. (JavadocVariableCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[118]: 'VARIABLE_DEF' should be separated from previous statement. (EmptyLineSeparatorCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[118]: Missing a Javadoc comment. (JavadocVariableCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[118]: Name 'rqMethod' must match pattern '^(id|[a-z]{3,12})$'. (MemberNameCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[119]: 'VARIABLE_DEF' should be separated from previous statement. (EmptyLineSeparatorCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[119]: Missing a Javadoc comment. (JavadocVariableCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[119]: Name 'rqHref' must match pattern '^(id|[a-z]{3,12})$'. (MemberNameCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[120]: 'VARIABLE_DEF' should be separated from previous statement. (EmptyLineSeparatorCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[120]: Missing a Javadoc comment. (JavadocVariableCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[120]: Name 'ms' must match pattern '^(id|[a-z]{3,12})$'. (MemberNameCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[122]: Line is longer than 80 characters (found 94). (LineLengthCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[122]: Missing a Javadoc comment. (JavadocMethodCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[122]: Redundant 'public' modifier. (RedundantModifierCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[122]: Parameter target should be final. (FinalParametersCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[122]: Parameter request should be final. (FinalParametersCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[122]: Parameter start should be final. (FinalParametersCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[130]: Parameter pattern should be final. (FinalParametersCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[130]: Parameter objects should be final. (FinalParametersCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[131]: Reference to instance variable 'logger' needs "this.". (RequireThisCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[132]: Line is longer than 80 characters (found 88). (LineLengthCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[132]: Name 'objectList' must match pattern '^(id|[a-z]{3,12})$'. (LocalFinalVariableNameCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[137]: Line is longer than 80 characters (found 86). (LineLengthCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[137]: Reference to instance variable 'logger' needs "this.". (RequireThisCheck)
[INFO] Checkstyle: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[137]: Concatenation of string literals prohibited (StringLiteralsConcatenationCheck)
[INFO] PMD: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[117-117]: The Logger variable declaration does not contain the static and final modifiers (LoggerIsNotStaticFinal)
[INFO] PMD: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[93-93]: Local variable 'log' could be declared final (LocalVariableCouldBeFinal)
[INFO] PMD: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[122-122]: Parameter 'target' is not assigned and could be declared final (MethodArgumentCouldBeFinal)
[INFO] PMD: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[122-122]: Parameter 'request' is not assigned and could be declared final (MethodArgumentCouldBeFinal)
[INFO] PMD: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[122-122]: Parameter 'start' is not assigned and could be declared final (MethodArgumentCouldBeFinal)
[INFO] PMD: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[130-130]: Parameter 'pattern' is not assigned and could be declared final (MethodArgumentCouldBeFinal)
[INFO] PMD: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[130-130]: Parameter 'objects' is not assigned and could be declared final (MethodArgumentCouldBeFinal)
[INFO] PMD: C:\projects\takes\src\main\java\org\takes\tk\TkSlf4j.java[137-137]: Missing arguments, expected 2 arguments but have 1 (InvalidSlf4jMessageFormat)
[INFO] Read our quality policy: http://www.qulice.com/quality.html
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:52 min
[INFO] Finished at: 2019-12-05T08:52:54+00:00
[INFO] Final Memory: 63M/1483M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.qulice:qulice-maven-plugin:0.18.5:check (jcabi-check) on project takes: Failure: There are 42 violations -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=2g; support was removed in 8.0
Command exited with code 1`

**Comment by fanifieiev** [comment]
@pal548 Just to make sure you understand the status of build, please watch the build status here every time you push new commit.
image

**Comment by fanifieiev** [comment]
@pal548 One of good design ideas to accomplish this task would be to break down the functionality into tiny objects and combine them after to reach the goal.
Think about an approach of having separate objects just for request and then for response as well which would log its own parts and then combine them in the `Take `.

**Comment by pal548** [comment]
@fanifieiev  oh, sorry, I didn't see your comments until now, I thought you are waiting for qulice to be ok)

File: src/main/java/org/takes/tk/TkSlf4j.java [hunk 1]
```diff
@@ -87,44 +90,52 @@ public TkSlf4j(final Take take, final String tgt) {
...
-                    new RqMethod.Base(req).method(),
-                    new RqHref.Base(req).href(),
-                    ex.getClass().getCanonicalName(),
-                    ex.getLocalizedMessage(),
-                    System.currentTimeMillis() - start
-                );
+                final List<Object> objectList = new ArrayList<>(2 + objects.length + 1);
+                objectList.add(this.rqMethod);
+                objectList.add(this.rqHref);
+                objectList.addAll(Arrays.asList(objects));
```
[the remark this answers is not in the document]
Comment (pal548) [reply src/main/java/org/takes/tk/TkSlf4j.java]: Arrays.asList() ? Is there something similar in cactoos?

**Comment by pal548** [comment]
> @pal548 One of good design ideas to accomplish this task would be to break down the functionality into tiny objects and combine them after to reach the goal.
> Think about an approach of having separate objects just for request and then for response as well which would log its own parts and then combine them in the `Take `.

Umm, thanks, I didn't notice, that first part of the log string is about request only, and other part is totally about the response, be it an ok response or exception.

Silly question, what is this `Take` ? Or maybe I can read about it somewhere?

File: src/main/java/org/takes/tk/TkSlf4j.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (fanifieiev) [reply src/main/java/org/takes/tk/TkSlf4j.java]: @pal548 I have already mentioned here, but will put it again:
`you need to start all the comments from the name of the person you are referring to`

[the remark this answers is not in the document]
Comment (fanifieiev) [reply src/main/java/org/takes/tk/TkSlf4j.java]: @pal548 Refer to`org.cactoos.list.ListOf`

**Comment by fanifieiev** [comment]
> @fanifieiev
> 
> > @pal548 One of good design ideas to accomplish this task would be to break down the functionality into tiny objects and combine them after to reach the goal.
> > Think about an approach of having separate objects just for request and then for response as well which would log its own parts and then combine them in the `Take `.
> 
> Umm, thanks, I didn't notice, that first part of the log string is about request only, and other part is totally about the response, be it an ok response or exception.
> 
> So for now please hold for a while, I'll try to make it as you suggested.

@pal548 Also, please leverage one of the existing interfaces or classes in Cactoos or Takes, try not to create any interfaces.

**Comment by fanifieiev** [comment]
@pal548 You do not need to create any extra interfaces or classes, use lambda expression with conjunction of Cactoos primitives like: 'org.cactoos.text.Joined', 'FormattedText' and 'Func'

**Comment by pal548** [comment]
@fanifieiev ok, I'll try it tomorrow

**Comment by fanifieiev** [comment]
> @fanifieiev ok, I'll try it tomorrow

@pal548 Ok, just bear in mind that the review may take up to 8 days only.
If there is no resolution, the PR will be declined and closed.

**Comment by pal548** [comment]
@fanifieiev Hello, I made it by 4 functions like Func<Request, String> and so on.
Now act() looks like that
```
            if (logger.isInfoEnabled()) {
                logger.info(
                    new Joined(" ",
                        funcRq.apply(req),
                        "returned",
                        funcRs.apply(rsp),
                        funcMs.apply(
                            System.currentTimeMillis() - start)
                    ).asString()
                );
```
Seems not very elegant... And I found only one use of Func<> in the whole project.
(I made it all, but there a bug in Qulice, see further)
What do you think about that design?

And I got the problem with Qulice - it fails with NPE on my file:
```
Caused by: java.lang.NullPointerException
    at com.qulice.checkstyle.ConstantUsageCheck.getText (ConstantUsageCheck.java:143)
    at com.qulice.checkstyle.ConstantUsageCheck.parseVarDef (ConstantUsageCheck.java:127)
    at com.qulice.checkstyle.ConstantUsageCheck.checkField (ConstantUsageCheck.java:87)
    at com.qulice.checkstyle.ConstantUsageCheck.visitToken (ConstantUsageCheck.java:68)
```
Is there a way to fix Qulice?

**Comment by fanifieiev** [comment]
@pal548 no need for 4 functions, make just one.

**Comment by fanifieiev** [comment]
@pal548 I also need to see full commit to understand why checkstyle fails

**Comment by pal548** [comment]
@fanifieiev I pushed the commits

**Comment by fanifieiev** [comment]
> @pal548 Make one function that would accept object of pattern message with params.
> That way you can always build any messages with any number of params.

@pal548 You can also create function inside method.

**Comment by fanifieiev** [comment]
@pal548 Your function should be used the following way:
```
function.apply(
   new MapEntry<>(
        "returned \"%s\"",
         new Object[]{rsp.head().iterator().next()}
   )
)
```
```
function.apply(
   new MapEntry<>(
        "thrown %s(\"%s\")",
        new Object[]{
           ex.getClass().getCanonicalName(),
           ex.getLocalizedMessage()
  }
)
)
```
That way it is flexible to build any message

**Comment by fanifieiev** [comment]
> And I found only one use of Func<> in the whole project.

@pal548 Regarding your message:`And I found only one use of Func<> in the whole project.`
You can find more examples in [Cactoos](https://github.com/yegor256/cactoos) project

**Comment by fanifieiev** [comment]
@pal548 Ok, use [**this**](https://github.com/fanifieiev/takes/commit/9724eebd1d2aee1f570606f317e0e9415ad15fc7) as an example

**Comment by pal548** [comment]
@fanifieiev Ok, thanks, I'll do it later today, or tomorrow.

**Comment by fanifieiev** [comment]
@paulodamaso Now it look good. Can you merge it, please?

**Comment by pal548** [comment]
@fanifieiev Oh, one more thing. Should we move logger creation from act() to constructor?

**Comment by fanifieiev** [comment]
> @fanifieiev Oh, one more thing. Should we move `logger` creation from `act()` to constructor? I tried to it, but there was a problem with lombok's `@ToString`

@pal548 This is at least out of scope of this task.
When doing some fixes/development of any tasks, the changes should reflect the original task only, so we are trying to develop here by means of small changes.
If there is sth you want to change, you need to leave a @todo with specific convention and if it is accepted by ARC when the PR is merged, it will be a separated task created by 0crat.

So, no need to do that now, as I already mentioned, there is an open discussion [here](https://github.com/yegor256/takes/issues/1004) regarding the logging.

**Comment by fanifieiev** [comment]
> @paulodamaso Now it looks good. Can you merge it, please?

@paulodamaso PING

**Comment by paulodamaso** [comment]
@fanifieiev Thanks for the review, and sorry for the delay

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1003](https://github.com/yegor256/takes/issues/1003) is now out of scope

**Comment by sereshqua** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by 0crat** [comment]
Quality review completed: +4 point(s) just awarded to @sereshqua[/z](https://www.0crat.com/u/sereshqua)

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev)

**Pull request merged**